File: backend/account-manager/src/main/java/com/jay/accountmanager/service/DashboardService.java
package com.jay.accountmanager.service;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.time.YearMonth;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.jay.accountmanager.dto.CategorySpendingDTO;
import com.jay.accountmanager.dto.DashboardResponseDTO;
import com.jay.accountmanager.dto.DashboardStatsDTO;
import com.jay.accountmanager.dto.MonthlySpendingDTO;
import com.jay.accountmanager.dto.PaymentDTO;
import com.jay.accountmanager.dto.TransactionDTO;
import com.jay.accountmanager.model.Payment;
import com.jay.accountmanager.model.Person;
import com.jay.accountmanager.model.Transaction;
import com.jay.accountmanager.model.TransactionType;
import com.jay.accountmanager.repository.AccountRepository;
import com.jay.accountmanager.repository.PaymentRepository;
import com.jay.accountmanager.repository.PersonRepository;
import com.jay.accountmanager.repository.TransactionRepository;
import com.jay.security.models.User;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class DashboardService {
    private final TransactionRepository transactionRepository;
    private final PaymentRepository paymentRepository;
    private final AccountRepository accountRepository;
    private final PersonRepository personRepository;

    private Person getPersonFromUser(User user) {
        if (user == null) {
            throw new SecurityException("No authenticated user found");
        }
        return personRepository.findByUserId(user.getId())
            .orElseThrow(() -> new SecurityException("Person not found for user: " + user.getUsername()));
    }

    @Transactional(readOnly = true)
    public DashboardResponseDTO getDashboardData(User user) {
        Person person = getPersonFromUser(user);
        LocalDateTime now = LocalDateTime.now();
        YearMonth currentMonth = YearMonth.from(now);
        LocalDateTime monthStart = currentMonth.atDay(1).atStartOfDay();
        LocalDateTime monthEnd = currentMonth.atEndOfMonth().atTime(23, 59, 59);

        return DashboardResponseDTO.builder()
            .stats(calculateStats(person, monthStart, monthEnd))
            .spendingByCategory(getSpendingByCategory(person, monthStart, monthEnd))
            .upcomingPayments(getUpcomingPayments(person, now))
            .recentTransactions(getRecentTransactions(person))
            .monthlyData(getMonthlyData(person))
            .build();
    }

    private DashboardStatsDTO calculateStats(Person person, LocalDateTime monthStart, LocalDateTime monthEnd) {
        // Get total balance across all accounts
        BigDecimal accountBalance = accountRepository.getTotalBalance(person);
        if (accountBalance == null) {
            accountBalance = BigDecimal.ZERO;
        }

        // Monthly Income
        BigDecimal monthlyIncome = transactionRepository.sumByPersonAndTypeAndDateBetween(
            person, TransactionType.INCOME, monthStart, monthEnd);

        // Monthly Bills
        BigDecimal monthlyBills = transactionRepository.sumByPersonAndTypeAndDateBetween(
            person, TransactionType.EXPENSE, monthStart, monthEnd);

        // Calculate savings progress
        double savingsProgress = 0.0;
        if (person.getSavingsGoal() != null && person.getSavingsGoal().compareTo(BigDecimal.ZERO) > 0) {
            savingsProgress = accountBalance.divide(person.getSavingsGoal(), 4, BigDecimal.ROUND_HALF_UP)
                .multiply(BigDecimal.valueOf(100))
                .doubleValue();
        }

        return DashboardStatsDTO.builder()
            .accountBalance(accountBalance)
            .monthlyIncome(monthlyIncome != null ? monthlyIncome : BigDecimal.ZERO)
            .monthlyBills(monthlyBills != null ? monthlyBills : BigDecimal.ZERO)
            .savingsGoal(person.getSavingsGoal())
            .savingsProgress(savingsProgress)
            .nextPaydate(calculateNextPaydate(person))
            .daysLeftInMonth(monthEnd.getDayOfMonth() - LocalDateTime.now().getDayOfMonth())
            .build();
    }

    private LocalDateTime calculateNextPaydate(Person person) {
        return transactionRepository.findLastIncomeDate(person)
            .map(lastDate -> {
                YearMonth nextMonth = YearMonth.from(lastDate).plusMonths(1);
                return nextMonth.atDay(lastDate.getDayOfMonth()).atStartOfDay();
            })
            .orElseGet(() -> LocalDateTime.now().plusDays(14));
    }

    private List<CategorySpendingDTO> getSpendingByCategory(Person person, LocalDateTime start, LocalDateTime end) {
        return transactionRepository.findSpendingByCategory(person, start, end);
    }

    private List<PaymentDTO> getUpcomingPayments(Person person, LocalDateTime now) {
        return paymentRepository.findUpcomingPaymentsByPerson(person, now, Pageable.ofSize(5)).stream()
            .map(payment -> PaymentDTO.builder()
                .id(payment.getId())
                .title(payment.getTitle())
                .amount(payment.getAmount())
                .dueDate(payment.getDueDate())
                .status(payment.getStatus())
                .recurrence(payment.getRecurrence())
                .description(payment.getDescription())
                .accountId(payment.getAccount().getId())
                .categoryId(payment.getCategory() != null ? payment.getCategory().getId() : null)
                .categoryName(payment.getCategory() != null ? payment.getCategory().getName() : null)
                .build())
            .collect(Collectors.toList());
    }

    private List<TransactionDTO> getRecentTransactions(Person person) {
        return transactionRepository.findRecentTransactionsByPerson(person, Pageable.ofSize(5)).stream()
            .map(transaction -> TransactionDTO.builder()
                .id(transaction.getId())
                .description(transaction.getDescription())
                .amount(transaction.getAmount())
                .date(transaction.getDate())
                .type(transaction.getType())
                .accountId(transaction.getAccount().getId())
                .categoryId(transaction.getCategory() != null ? transaction.getCategory().getId() : null)
                .categoryName(transaction.getCategory() != null ? transaction.getCategory().getName() : null)
                .notes(transaction.getNotes())
                .build())
            .collect(Collectors.toList());
    }

    private List<MonthlySpendingDTO> getMonthlyData(Person person) {
        List<MonthlySpendingDTO> monthlyData = new ArrayList<>();
        YearMonth current = YearMonth.now();

        for (int i = 5; i >= 0; i--) {
            YearMonth month = current.minusMonths(i);
            LocalDateTime start = month.atDay(1).atStartOfDay();
            LocalDateTime end = month.atEndOfMonth().atTime(23, 59, 59);

            BigDecimal income = transactionRepository.sumByPersonAndTypeAndDateBetween(
                person, TransactionType.INCOME, start, end);
            BigDecimal bills = transactionRepository.sumByPersonAndTypeAndDateBetween(
                person, TransactionType.EXPENSE, start, end);

            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("MMM yyyy");
            String formattedMonth = month.format(formatter);

            monthlyData.add(MonthlySpendingDTO.builder()
                .month(formattedMonth)
                .income(income != null ? income : BigDecimal.ZERO)
                .bills(bills != null ? bills : BigDecimal.ZERO)
                .build());
        }

        return monthlyData;
    }

    @Transactional(readOnly = true)
    public Page<TransactionDTO> getTransactions(User user, Pageable pageable) {
        Person person = getPersonFromUser(user);
        Page<Transaction> transactions = transactionRepository.findByAccountPerson(person, pageable);
        
        List<TransactionDTO> dtos = transactions.getContent().stream()
            .map(transaction -> TransactionDTO.builder()
                .id(transaction.getId())
                .description(transaction.getDescription())
                .amount(transaction.getAmount())
                .date(transaction.getDate())
                .type(transaction.getType())
                .accountId(transaction.getAccount().getId())
                .categoryId(transaction.getCategory() != null ? transaction.getCategory().getId() : null)
                .categoryName(transaction.getCategory() != null ? transaction.getCategory().getName() : null)
                .notes(transaction.getNotes())
                .build())
            .collect(Collectors.toList());

        return new PageImpl<>(dtos, pageable, transactions.getTotalElements());
    }

    @Transactional(readOnly = true)
    public Page<PaymentDTO> getPayments(User user, Pageable pageable) {
        Person person = getPersonFromUser(user);
        Page<Payment> payments = paymentRepository.findByAccountPerson(person, pageable);
        
        List<PaymentDTO> dtos = payments.getContent().stream()
            .map(payment -> PaymentDTO.builder()
                .id(payment.getId())
                .title(payment.getTitle())
                .amount(payment.getAmount())
                .dueDate(payment.getDueDate())
                .status(payment.getStatus())
                .recurrence(payment.getRecurrence())
                .description(payment.getDescription())
                .accountId(payment.getAccount().getId())
                .categoryId(payment.getCategory() != null ? payment.getCategory().getId() : null)
                .categoryName(payment.getCategory() != null ? payment.getCategory().getName() : null)
                .build())
            .collect(Collectors.toList());

        return new PageImpl<>(dtos, pageable, payments.getTotalElements());
    }
}

File: frontend/src/services/dashboard.service.ts
import type { DashboardData, Transaction, Payment } from '@/types/dashboard'
import { apiService } from '@/services/api.service'
import type { AxiosResponse } from 'axios'

interface TransactionResponse {
  transactions: Transaction[]
  total: number
}

class DashboardService {
  private readonly BASE_URL = '/api/dashboard'

  async getDashboardData(): Promise<DashboardData> {
    const response: AxiosResponse<DashboardData> = await apiService.get(`${this.BASE_URL}`)
    if (!response || !('data' in response)) {
      throw new Error('Invalid response format')
    }
    return this.transformDashboardData(response.data)
  }

  async getTransactions(params: {
    page?: number
    limit?: number
    startDate?: Date
    endDate?: Date
    type?: 'income' | 'expense'
    categoryId?: string
  }): Promise<{ transactions: Transaction[]; total: number }> {
    const response: AxiosResponse<TransactionResponse> = await apiService.get(`${this.BASE_URL}/transactions`, { params })
    if (!response || !('data' in response)) {
      throw new Error('Invalid response format')
    }
    return {
      transactions: response.data.transactions.map(this.transformTransaction),
      total: response.data.total
    }
  }

  async getUpcomingPayments(params: {
    limit?: number
    startDate?: Date
    endDate?: Date
  }): Promise<Payment[]> {
    const response: AxiosResponse<Payment[]> = await apiService.get(`${this.BASE_URL}/upcoming-payments`, { params })
    if (!response || !('data' in response)) {
      throw new Error('Invalid response format')
    }
    return response.data.map(this.transformPayment)
  }

  async addTransaction(transaction: Omit<Transaction, 'id' | 'date'>): Promise<Transaction> {
    const response: AxiosResponse<Transaction> = await apiService.post(`${this.BASE_URL}/transactions`, transaction)
    if (!response || !('data' in response)) {
      throw new Error('Invalid response format')
    }
    return this.transformTransaction(response.data)
  }

  async updateTransaction(id: string, transaction: Partial<Transaction>): Promise<Transaction> {
    const response: AxiosResponse<Transaction> = await apiService.put(`${this.BASE_URL}/transactions/${id}`, transaction)
    if (!response || !('data' in response)) {
      throw new Error('Invalid response format')
    }
    return this.transformTransaction(response.data)
  }

  async deleteTransaction(id: string): Promise<void> {
    await apiService.delete(`${this.BASE_URL}/transactions/${id}`)
  }

  async addPayment(payment: Omit<Payment, 'id'>): Promise<Payment> {
    const response: AxiosResponse<Payment> = await apiService.post(`${this.BASE_URL}/upcoming-payments`, payment)
    if (!response || !('data' in response)) {
      throw new Error('Invalid response format')
    }
    return this.transformPayment(response.data)
  }

  async updatePayment(id: string, payment: Partial<Payment>): Promise<Payment> {
    const response: AxiosResponse<Payment> = await apiService.put(`${this.BASE_URL}/upcoming-payments/${id}`, payment)
    if (!response || !('data' in response)) {
      throw new Error('Invalid response format')
    }
    return this.transformPayment(response.data)
  }

  async deletePayment(id: string): Promise<void> {
    await apiService.delete(`${this.BASE_URL}/upcoming-payments/${id}`)
  }

  private transformDashboardData(data: any): DashboardData {
    if (!data) {
      throw new Error('Dashboard data is undefined');
    }

    return {
      stats: {
        totalIncome: Number(data.stats?.totalIncome || 0),
        totalExpenses: Number(data.stats?.totalExpenses || 0),
        balance: Number(data.stats?.balance || 0),
        monthlyIncome: Number(data.stats?.monthlyIncome || 0),
        monthlyExpenses: Number(data.stats?.monthlyExpenses || 0),
        monthlyBalance: Number(data.stats?.monthlyBalance || 0),
        savingsGoal: Number(data.stats?.savingsGoal || 0),
        savingsProgress: Number(data.stats?.savingsProgress || 0),
        ...(data.stats?.nextPaydate && { nextPaydate: new Date(data.stats.nextPaydate) })
      },
      upcomingPayments: Array.isArray(data.upcomingPayments) 
        ? data.upcomingPayments.map(this.transformPayment)
        : [],
      recentTransactions: Array.isArray(data.recentTransactions)
        ? data.recentTransactions.map(this.transformTransaction)
        : [],
      spendingByCategory: Array.isArray(data.spendingByCategory)
        ? data.spendingByCategory.map((category: any) => ({
            name: category.name || '',
            amount: Number(category.amount || 0),
            color: category.color,
            icon: category.icon
          }))
        : [],
      monthlyFinancials: Array.isArray(data.monthlyFinancials)
        ? data.monthlyFinancials.map((month: any) => ({
            month: month.month || '',
            income: Number(month.income || 0),
            expenses: Number(month.expenses || 0)
          }))
        : []
    };
  }

  private transformTransaction(transaction: any): Transaction {
    return {
      id: transaction.id,
      title: transaction.title,
      amount: transaction.amount,
      date: new Date(transaction.date),
      type: transaction.type,
      category: {
        id: transaction.category.id,
        name: transaction.category.name,
        color: transaction.category.color,
        icon: transaction.category.icon
      },
      ...(transaction.description && { description: transaction.description })
    }
  }

  private transformPayment(payment: any): Payment {
    return {
      id: payment.id,
      title: payment.title,
      description: payment.description,
      amount: payment.amount,
      dueDate: new Date(payment.dueDate),
      category: {
        id: payment.category.id,
        name: payment.category.name,
        color: payment.category.color,
        icon: payment.category.icon
      },
      status: payment.status,
      ...(payment.recurrence && { recurrence: payment.recurrence })
    }
  }
}

export const dashboardService = new DashboardService()
